src/main/java/emissary/util/io/LoggingPrintStream.java

```diff
@@ -0,0 +1,172 @@
+     * @param closeWaitTimeUnit the time unit to use for the closeWaitTime.
+     */
+    public LoggingPrintStream(final OutputStream outputStream, final String streamName, final Logger logger, final Level level,
+            final long closeWaitTime, final TimeUnit closeWaitTimeUnit) {
+        super(outputStream);
+
+        Validate.notNull(streamName, "Required: streamName not null");
+        Validate.notNull(logger, "Required: logger not null");
+        Validate.notNull(level, "Required: level not null");
+        Validate.isTrue(closeWaitTime >= 0, "Required: closeWaitTime >= 0");
+        Validate.notNull(closeWaitTimeUnit, "Required: closeWaitTimeUnit not null");
+
+        this.streamName = streamName;
+        this.logger = logger;
+        this.level = level;
+        this.closeWaitTime = closeWaitTime;
+        this.closeWaitTimeUnit = closeWaitTimeUnit;
+        this.executorService = Executors.newSingleThreadExecutor(
+                runnable -> new Thread(runnable, "LoggingPrintStream_LoggingThread"));
+
+    }
+
+    @Override
+    public void print(final String string) {
+        if (linesToSkip == 0) {
+            executorService.submit(() -> log(logger, level, NORMAL_LOG_FORMAT, streamName, string));
+        }
+
+        super.print(string);
+    }
+
+    @Override
+    public void println(final Object object) {
+        if (object instanceof Throwable) {
+            final Throwable throwable = (Throwable) object;
+            final StringWriter stringWriter = new StringWriter();
+            final PrintWriter printWriter = new PrintWriter(stringWriter);
+
+            throwable.printStackTrace(printWriter);
+
+            final String throwableString = stringWriter.toString();
+
+            try (StringReader stringReader = new StringReader(throwableString);
+                    LineNumberReader lineNumberReader = new LineNumberReader(stringReader)) {
+                lineNumberReader.skip(Long.MAX_VALUE);
+                linesToSkip = lineNumberReader.getLineNumber();
+            } catch (IOException e) {
+                logger.error("This should never happen as the Readers are backed by a String.");
+            }
+
+            executorService.submit(() -> log(logger, level, THROWABLE_LOG_FORMAT, streamName, throwable));
+        }
+
+        super.println(object);
+
+        if (linesToSkip > 0) {
+            linesToSkip--;
+        }
+    }
+
+    @Override
+    public void close() {
+        try {
+            if (!executorService.isShutdown()) {
+                executorService.shutdown();
+                executorService.awaitTermination(closeWaitTime, closeWaitTimeUnit);
+            }
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+        }
+
+        super.close();
+    }
+
+    private static void log(final Logger logger, final Level level, final String string, final Object object1, final Object object2) {
```

Comment:
If we remove the `static` modifier, we can access the existing `logger` and `level` fields directly

```suggestion
    private void log( final String string, final Object object1, final Object object2, final Map<String, String> ctx) {
        try {
            ctx.forEach((k,v) -> MDC.put(k,v));
            if (ctx != null) {
                ctx.forEach((k,v) -> MDC.put(k,v));
            }
            ...
        } finally {
            MDC.clear();
        }
```